go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -419,8 +420,30 @@ func (tm *TabletManager) demotePrimary(ctx context.Context, revertPartialFailure
 		if err := tm.QueryServiceControl.SetServingType(tablet.Type, logutil.ProtoToTime(tablet.PrimaryTermStartTime), false, "demotion in progress"); err != nil {
 			return nil, vterrors.Wrap(err, "SetServingType(serving=false) failed")
 		}
+
+		// Note that due to an upstream semi-sync bug the semi-sync ACK request may be lost, causing the
+		// current primary to wait forever for the ACK. See:
+		//   https://bugs.mysql.com/bug.php?id=104015
+		//   https://bugs.mysql.com/bug.php?id=101056
+		// This in turns causes the statement to disable [Super]ReadOnly mode to block on a mutex that
+		// the semi-sync plugin is holding while waiting for the ACK. This then can cause the shard to
+		// be left in an unhealthy state without a serving PRIMARY. So here we need to disable SemiSync
+		// which will ensure that we are unblocked by this scenario (tablet repair should also fix this
+		// as needed via fixSemiSync()).
+		if wasSemiSyncSourceEnabled {
+			log.Infof("DemotePrimary disabling source semi-sync")
+			if err := tm.MysqlDaemon.SetSemiSyncEnabled(false, wasSemiSyncReplicaEnabled); err != nil {
+				return nil, vterrors.Wrapf(err, "SetSemiSyncEnabled(source=false, replica=%v) failed", wasSemiSyncReplicaEnabled)
+			}
```

Comment:
I don't think we need these changes, we have code to stop the semi-sync primary side just below the set read only part. We should just reorder them, because this code makes it so, that that code will not run

Reply:
I have created a PR for these changes - https://github.com/vitessio/vitess/pull/10383. WDYT @deepthi and @mattlord about reordering the statements instead of adding these which essentially accomplish the same things?

Reply:
Sounds good! I’ll close this one

Reply:
The perils of looking at GH diffs without looking at the rest of the file 😭

Reply:
I know right. I knew that we had code for semi-sync changes down below because I worked with it for the semi-sync changes for vtorc